### PR TITLE
PLAT-471: test pending counters and earliest pulses in VStateReport

### PR DESCRIPTION
**Pull request by ruz**
table test that covers many variouse situations

### CODE REVIEW

File: ledger-core/virtual/integration/vstatereport_test.go [hunk 1]
```diff
@@ -92,3 +99,368 @@ func TestVirtual_SendVStateReport_IfPulseChanged(t *testing.T) {
 
 	require.Equal(t, 1, countVStateReport)
 }
+
+func TestVirtual_StateReport_CheckPendingCountersAndPulses(t *testing.T) {
+	t.Log("C FIXME ")
```
Comment (aalekseev-ru) [inline ledger-core/virtual/integration/vstatereport_test.go:104]: I don't know all test naming rules, but is it ok? @ElSacramento

Comment (MariaVasilenko) [reply ledger-core/virtual/integration/vstatereport_test.go:104]: C5111

File: ledger-core/virtual/integration/vstatereport_test.go [hunk 2]
```diff
@@ -92,3 +100,502 @@ func TestVirtual_SendVStateReport_IfPulseChanged(t *testing.T) {
...
+	class := gen.UniqueReference()
+
+	for _, test := range table {
+		t.Run(test.name, func(t *testing.T) {
+
+			suite := &stateReportCheckPendingCountersAndPulsesTest{}
+			ctx := suite.initServer(t)
+			defer suite.finish()
+
+			suite.createPulsesFromP1toP4(ctx)
```
Comment (bigbes) [inline ledger-core/virtual/integration/vstatereport_test.go:229]: unify creation of pulses:
```
func (s *stateReportCheckPendingCountersAndPulsesTest) incrementPulses(ctx context.Context, count int) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	for i := 0; i < count; i++ {
		s.server.IncrementPulseAndWaitIdle(ctx)
		s.pulses[s.currentPulseIndex] = s.server.GetPulse().PulseNumber
		s.currentPulseIndex++
	}
}
```

File: ledger-core/virtual/integration/vstatereport_test.go [hunk 3]
```diff
@@ -92,3 +100,502 @@ func TestVirtual_SendVStateReport_IfPulseChanged(t *testing.T) {
...
+
+type stateReportCheckPendingCountersAndPulsesTest struct {
+	mu sync.RWMutex
+
+	mc         *minimock.Controller
+	server     *utils.Server
+	runnerMock *logicless.ServiceMock
+
+	currentPulseIndex int
+	pulses            [6]pulse.Number
```
Comment (bigbes) [inline ledger-core/virtual/integration/vstatereport_test.go:126]: why len is six and zero is unused?

File: ledger-core/virtual/integration/vstatereport_test.go [hunk 4]
```diff
@@ -92,3 +100,502 @@ func TestVirtual_SendVStateReport_IfPulseChanged(t *testing.T) {
...
+	msg := utils.NewRequestWrapper(s.getCurrentPulse(), pl).SetSender(me).Finalize()
+	s.server.SendMessage(ctx, msg)
+}
+
+func (s *stateReportCheckPendingCountersAndPulsesTest) addPayloadAndWaitIdle(
+	ctx context.Context, pl payload.Marshaler,
+) {
+	s.server.SuspendConveyorAndWaitThenResetActive()
+	s.addPayload(ctx, pl)
+	s.server.WaitActiveThenIdleConveyor()
```
Comment (bigbes) [inline ledger-core/virtual/integration/vstatereport_test.go:595]: why?

Comment (ruz) [reply ledger-core/virtual/integration/vstatereport_test.go:595]: to get deterministic order of operations as much as I can

File: ledger-core/virtual/integration/vstatereport_test.go (hunk 3 above)
Comment (ruz) [reply ledger-core/virtual/integration/vstatereport_test.go:126]: zero is used for "zero pulse number" (empty) and used in checks where needed

File: ledger-core/virtual/integration/vstatereport_test.go (hunk 2 above)
Comment (ruz) [reply ledger-core/virtual/integration/vstatereport_test.go:236]: discussed this

File: ledger-core/virtual/integration/vstatereport_test.go (hunk 3 above)
Comment (ruz) [reply ledger-core/virtual/integration/vstatereport_test.go:126]: added as many info as I could around value under index 0 in this array

File: ledger-core/virtual/integration/vstatereport_test.go (hunk 4 above)
Comment (ruz) [reply ledger-core/virtual/integration/vstatereport_test.go:609]: discussed this

**Pull request merged**